shell/platform/darwin/macos/framework/Source/FlutterTextInputSemanticsObject.mm

```diff
@@ -0,0 +1,212 @@
+  self = [super initWithFrame:NSZeroRect];
+  if (self) {
+    _node = node;
+    [self setCell:[[FlutterTextFieldCell alloc] initWithTextField:self]];
+  }
+  return self;
+}
+
+- (void)setFieldEditor:(FlutterTextInputPlugin*)plugin {
+  _plugin = plugin;
+  NSAssert([self.cell isKindOfClass:[FlutterTextFieldCell class]],
+           @"The cell of a FlutterTextField must be a FlutterTextFieldCell");
+  FlutterTextFieldCell* cell = (FlutterTextFieldCell*)self.cell;
+  [cell setFieldEditor:plugin];
+}
+
+- (void)updateString:(NSString*)string withSelection:(NSRange)selection {
+  NSAssert(_plugin.client == self,
+           @"Can't update FlutterTextField when it is not the first responder");
+  if (![[self stringValue] isEqualToString:string]) {
+    [self setStringValue:string];
+  }
+  if (!NSEqualRanges(_plugin.selectedRange, selection)) {
+    [_plugin setSelectedRange:selection];
+  }
+}
+
+#pragma mark - NSView
+
+- (NSRect)frame {
+  return _node->GetFrameRelativeToControllerView();
+}
+
+#pragma mark - NSAccessibilityProtocol
+
+- (void)setAccessibilityFocused:(BOOL)isFocused {
+  [super setAccessibilityFocused:isFocused];
+  ui::AXActionData data;
+  data.action = isFocused ? ax::mojom::Action::kFocus : ax::mojom::Action::kBlur;
+  _node->GetDelegate()->AccessibilityPerformAction(data);
+}
+
+#pragma mark - NSResponder
+
+- (BOOL)becomeFirstResponder {
+  if (_plugin.client == self && [_plugin isFirstResponder]) {
+    // This text field is already the first responder.
+    return YES;
+  }
+  BOOL result = [super becomeFirstResponder];
+  if (result) {
+    _plugin.client = self;
+    // The default implementation of the becomeFirstResponder will change the
+    // text editing state. Need to manually set it back.
+    NSString* textValue = @(_node->GetStringAttribute(ax::mojom::StringAttribute::kValue).data());
+    int start = _node->GetIntAttribute(ax::mojom::IntAttribute::kTextSelStart);
+    int end = _node->GetIntAttribute(ax::mojom::IntAttribute::kTextSelEnd);
+    NSRange selection;
+    if (start > 0 && end > 0) {
+      selection = NSMakeRange(start, end - start);
+    } else {
+      selection = NSMakeRange([self stringValue].length, 0);
+    }
+    [self updateString:textValue withSelection:selection];
+  }
+  return result;
+}
+
+@end
+
+namespace flutter {
+
+FlutterTextPlatformNode::FlutterTextPlatformNode(FlutterPlatformNodeDelegate* delegate,
+                                                 FlutterEngine* engine) {
+  Init(delegate);
+  engine_ = engine;
+  native_text_field_ = [[FlutterTextField alloc] initWithPlatformNode:this];
+  [native_text_field_ setFieldEditor:engine.viewController.textInputPlugin];
+  native_text_field_.bezeled = NO;
+  native_text_field_.drawsBackground = NO;
+  native_text_field_.bordered = NO;
+  native_text_field_.focusRingType = NSFocusRingTypeNone;
+}
+
+FlutterTextPlatformNode::~FlutterTextPlatformNode() {
+  [native_text_field_ setFieldEditor:nil];
```

Comment:
Could you give an overview of the ownership relation between the various classes involved here? What object ultimately own the `FlutterTextPlatformNode`? It's not at all obvious to me what guarantee you have that `engine.viewController.textInputPlugin` will outlive this object instance.

Reply:
![image](https://user-images.githubusercontent.com/47866232/118702628-06c85080-b7ca-11eb-9941-c1be44750b2c.png)

You are right, there is no guarantee. I made the change to store the pointer to the engine directly and access the textinputplugin directly from the engine pointer, so it will either be nil or a valid pointer.